File: 0300/dp.cpp
#include <gtest/gtest.h>

#include <algorithm>
#include <vector>

using namespace std;

// time complexity: O(N^2)
// space complexity: O(N)
class Solution {
 public:
  int lengthOfLIS(vector<int>& nums) {
    vector<int> length(nums.size(), 1);
    int ret = 1;
    for (size_t i = 1; i < nums.size(); ++i) {
      for (size_t j = 0; j < i; ++j) {
        if (nums[j] >= nums[i]) {
          continue;
        }
        length[i] = max(length[j] + 1, length[i]);
      }
      ret = max(length[i], ret);
    }
    return ret;
  }
};

int main() {
  Solution sol;
  vector<int> nums;
  int expected;

  nums = {10, 9, 2, 5, 3, 7, 101, 18};
  expected = 4;
  EXPECT_EQ(expected, sol.lengthOfLIS(nums));

  nums = {0};
  expected = 1;
  EXPECT_EQ(expected, sol.lengthOfLIS(nums));

  nums = {5, 4, 3, 2, 1};
  expected = 1;
  EXPECT_EQ(expected, sol.lengthOfLIS(nums));

  nums = {6, 7, 8, 9, 1, 1, 2, 3, 4, 5};
  expected = 5;
  EXPECT_EQ(expected, sol.lengthOfLIS(nums));
}

File: 0300/dp.rs
struct Solution;

// dp[i] is the length of longest increasing substring
// of all increasing substring whose last element is nums[i]
//  nums: [4, 2, 3, 1, 5]
//  dp:   [1, 1, 2, 1, 3]
// time complexity: O(n^2)
// space complexity: O(n)
impl Solution {
    pub fn length_of_lis(nums: Vec<i32>) -> i32 {
        let size: usize = nums.len();
        if size == 0 { return 0; }
        let mut ret: i32 = 1;
        let mut dp: Vec<i32> = vec![1; size];
        for idx in 1..size {
            for subidx in 0..idx {
                if nums[subidx] >= nums[idx] { continue; }
                dp[idx] = std::cmp::max(dp[subidx] + 1, dp[idx]);
            }
            ret = std::cmp::max(ret, dp[idx]);
        }
        return ret;
    }
}

fn main() {
    let mut nums;

    nums = vec![4];
    assert!(Solution::length_of_lis(nums) == 1);

    nums = vec![2,2];
    assert!(Solution::length_of_lis(nums) == 1);

    nums = vec![4,2,3,1,5];
    assert!(Solution::length_of_lis(nums) == 3);

    nums = vec![10,9,2,5,3,7,101,18];
    assert!(Solution::length_of_lis(nums) == 4);
}
